impeller/renderer/backend/vulkan/descriptor_pool_vk.cc

```diff
@@ -19,79 +18,58 @@ DescriptorPoolVK::DescriptorPoolVK(
 DescriptorPoolVK::~DescriptorPoolVK() = default;
 
 static vk::UniqueDescriptorPool CreatePool(const vk::Device& device,
-                                           uint32_t pool_count) {
+                                           uint32_t image_count,
+                                           uint32_t buffer_count) {
   TRACE_EVENT0("impeller", "CreateDescriptorPool");
-  std::vector<vk::DescriptorPoolSize> pools = {
-      {vk::DescriptorType::eCombinedImageSampler, pool_count},
-      {vk::DescriptorType::eUniformBuffer, pool_count},
-      {vk::DescriptorType::eStorageBuffer, pool_count}};
-
+  std::vector<vk::DescriptorPoolSize> pools = {};
+  if (image_count > 0) {
+    pools.emplace_back(vk::DescriptorPoolSize{
+        vk::DescriptorType::eCombinedImageSampler, image_count});
+  }
+  if (buffer_count > 0) {
+    pools.emplace_back(vk::DescriptorPoolSize{
+        vk::DescriptorType::eUniformBuffer, buffer_count});
+    pools.emplace_back(vk::DescriptorPoolSize{
+        vk::DescriptorType::eStorageBuffer, buffer_count});
+  }
   vk::DescriptorPoolCreateInfo pool_info;
-  pool_info.setMaxSets(pools.size() * pool_count);
+  pool_info.setMaxSets(image_count + buffer_count);
   pool_info.setPoolSizes(pools);
-
   auto [result, pool] = device.createDescriptorPoolUnique(pool_info);
   if (result != vk::Result::eSuccess) {
     VALIDATION_LOG << "Unable to create a descriptor pool";
   }
   return std::move(pool);
 }
 
-std::optional<vk::DescriptorSet> DescriptorPoolVK::AllocateDescriptorSet(
-    const vk::DescriptorSetLayout& layout,
-    size_t command_count) {
-  if (pools_.empty()) {
-    pool_size_ = command_count;
+std::vector<vk::DescriptorSet> DescriptorPoolVK::AllocateDescriptorSets(
```

Comment:
This weakens the error reporting ability of this API though. An empty vector isn't the same as a vector with nullopt elements.

Reply:
Yeah, I think this should be something like a status_or

Reply:
I updated these to status_or. not totally in love with the existing status enums, and the messages feel a bit redundant witht the validation logs.